### PR TITLE
Adds a Relational Memory Core cell

### PR DESCRIPTION
**Pull request by gugarosa**
# Description

Relational Memory Core networks recently gain the spotlight as they use attention-based mechanisms to improve memory learning in recurrent networks. According to Santoro et al. [1], they have outperformed LSTMs and GRUs in several tasks and decreased the performance gap between recurrent-based networks and transformer-based architectures.

Nevertheless, it is hard to find any implementation into well-known frameworks, such as Tensorflow and PyTorch. Thus, this pull request aims at fulfilling such a gap by adding the structure of a Relational Memory Core cell.

The files have been implemented accordingly to the contribution guidelines, as well as to the RNN-based guidelines. Additionally, throughout the implementation process, every function was checked and revised to match the style that has been proposed by other files in the `rnn` package. Finally, several tests to attest their shapes and calculation have been implemented, as well as all the contribution guideline scripts have been run, e.g., style, sanity check, and CPU/GPU-based tests.

[1] A. Santoro, et al. "Relational recurrent neural networks". Advances in neural information processing systems, 2018.

## Type of change

- [ ] Bug fix
- [ ] New Tutorial
- [ ] Updated or additional documentation
- [ ] Additional Testing
- [ ] New Activation and the changes conform to the [activation contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/activations/README.md#contribution-guidelines)
- [ ] New Callback and the changes conform to the [callback contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/callbacks/README.md#contribution-guidelines)
- [ ] New Image addition and the changes conform to the [image op contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/image/README.md#contribution-guidelines)
- [ ] New Layer and the changes conform to the [layer contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/layers/README.md#contribution-guidelines)
- [ ] New Loss and the changes conform to the [loss contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/losses/README.md#contribution-guidelines)
- [ ] New Metric and the changes conform to the [metric contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/metrics/README.md#contribution-guidelines)
- [ ] New Optimizer and the changes conform to the [optimizer contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/optimizers/README.md#contribution-guidelines)
- [X] New RNN Cell and the changes conform to the [rnn contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/rnn/README.md#contribution-guidelines)
- [ ] New Seq2seq addition and the changes conform to the [seq2seq contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/seq2seq/README.md#contribution-guidelines)
- [ ] New Text addition and the changes conform to the [text op contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/text/README.md#contribution-guidelines)

# Checklist:

- [X] I've properly [formatted my code according to the guidelines](https://github.com/tensorflow/addons/blob/master/CONTRIBUTING.md#coding-style)
    - [X] By running Black + Flake8
    - [ ] By running pre-commit hooks
- [ ] This PR addresses an already submitted issue for TensorFlow Addons
- [X] I have made corresponding changes to the documentation
- [X] I have added tests that prove my fix is effective or that my feature works
- [ ] This PR contains modifications to C++ custom-ops

# How Has This Been Tested?

If you're adding a bugfix or new feature please describe the tests that you ran to verify your changes:
*  `test_base_rmc`: tests whether the expected output and hidden state matches a small level of tolerance and the corresponding shapes;
*  `test_attend_over_memory_rmc`: tests whether the attended memory function is capable of receiving pre-defined inputs and producing correctly-shaped outputs;
* `test_keras_rnn`: tests whether an RMC-based cell can be used within the RNN class provided by Keras;
* `test_config_rmc`: tests whether the configuration object saved by the RMC can be serialized and de-serialized.

### CODE REVIEW

**Comment by bhack** [comment]
Thanks, I put this in ecosystem review. 
A this was mainly a Deepmind paper it had a sort of "official" implementation in Deepmind Sonnet at: https://github.com/deepmind/sonnet/blob/master/sonnet/python/modules/relational_memory.py
But it seems that Deepmind has not anymore this in Sonnet v2. /cc @tomerk @dynamicwebpaige I don't know if you have some more info about this or an interested review resource that worked with the original implementation/paper.

**Comment by gugarosa** [comment]
> Thanks, I put this in ecosystem review.
> A this was mainly a Deepmind paper it had a sort of "official" implementation in Deepmind Sonnet at: https://github.com/deepmind/sonnet/blob/master/sonnet/python/modules/relational_memory.py
> But it seems that Deepmind has not anymore this in Sonnet v2. /cc @tomerk @dynamicwebpaige I don't know if you have some more info about this or an interested review resource that worked with the original implementation/paper.

Perfect, thank you very much!

A while ago, I had the opportunity to glance at Sonnet, but preferred to work with TF as it provides a lower abstraction level. Thus, I ended up re-implementing it due to my thesis' ecosystem being on top of TF and because I did not want to add an extra requirement to the project.

Nevertheless, I am looking forward for any updates. If you need anything at all, please just let me know.

Best regards,
Gustavo.
